app/scripts/models/auth_brokers/web-channel.js

```diff
@@ -54,6 +54,55 @@ define([
       return p();
     },
 
+    // WebChannel reliers can request access to relier-specific encryption
+    // keys.  In the future this logic may be lifted into the base OAuth class
+    // and made available to all reliers, but we're putting it in this subclass
+    // for now to guard against accidental exposure.
+
+    getOAuthResult: function (account) {
+      var self = this;
+      return OAuthAuthenticationBroker.prototype.getOAuthResult.call(this, account)
+        .then(function (result) {
+          if (! self.relier.isKeyFetchEnabled()) {
+            return result;
+          }
+          return self.deriveRelierKeys(account)
+            .then(function (keys) {
+              result.keys = keys;
+              return result;
+            });
+        });
+    },
+
+    deriveRelierKeys: function (account) {
+      var self = this;
+      // XXX TODO: pass an fxaClient around properly
+      var fxaClient = this._assertionLibrary._fxaClient;
+      var masterKeys;
+      var relierKeys = {};
+      var infoPrefix = 'identity.mozilla.com/picl/v1/oauth/';
+      var keyFetchToken = account.get('keyFetchToken');
+      var unwrapBKey = account.get('unwrapBKey');
+      if (! keyFetchToken || ! unwrapBKey) {
+        return p(null);
+      }
+      return fxaClient.accountKeys(keyFetchToken, unwrapBKey)
+        .then(function (keys) {
+          masterKeys = keys;
+          var relierInfoA = infoPrefix + 'kAr:' + self.relier.get('clientId');
+          return fxaClient.generateDerivedKey(masterKeys.kA, 64, relierInfoA);
```

Comment:
What about creating higher level functions like `generateDerivedKA` and `generateDerivedKB` that accept a `kA` or `kB` and a `clientId`? It might make the code easier to follow.